---
title: "Untitled"
author: "Steindl S.,"
date: "10 9 2024"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(factoextra)
library(FactoMineR)
library(ggplot2)
library(ggrepel)
library(dplyr)
library(corrplot)
library(vegan)
library(plotly)
Rasdman <- read.csv("/media/inter/ssteindl/FC/usecaserepo/SYNC0524/uc3-drosophola-genetics/projects/WormPickerOOP/use_case/rasdaman_data.csv", header=T)

metafile <- read.csv("/media/inter/ssteindl/FC/usecaserepo/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/data/samps_25Feb2023.csv")
columns_with_only_na_string <- sapply(Rasdman, function(x) all(x == "na"))
# Display column names that contain only NA values
na_string_columns <- names(df)[columns_with_only_na_string]
Rclean <- Rasdman[, colSums(Rasdman == "na") != nrow(Rasdman)]
Rclean <- as.data.frame(Rclean[2:ncol(Rclean)])

df_numeric <- as.data.frame(lapply(Rclean, as.numeric))



```


## Preparing the Data
``` {r clean_data,fig.width=20, fig.height=14, echo=FALSE}

###Vienna

###Vienna S4E Data
ViennaData_LN <- read.csv("/media/inter/ssteindl/FC/usecaserepo/SYNC0524/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/S3/ViennaData/fairicube/vienna_data/100m/fairicube_merged_data_temp.csv")

ViennaData <- read.csv("/media/inter/ssteindl/FC/usecaserepo/SYNC0524/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/S3/ViennaData/fairicube/vienna_data/100m/ViennaPointData_UD.csv")

sampleMeta <- read.csv("/media/inter/ssteindl/FC/usecaserepo/SYNC0524/uc3-drosophola-genetics/projects/ViennaCityFly2024/ViennaCityFlySamples3009.csv")

fullInfo <- merge(sampleMeta, ViennaData, by="sampleId")

###Response Variables Selected Manually
responseSpeciesComposition <- as.data.frame(fullInfo[13:25])

#responseSpeciesComposition <- as.data.frame(cbind(fullInfo$melanogaster, fullInfo$simulans, fullInfo$suzukii, fullInfo$busckii, fullInfo$testacea, fullInfo$hydei, fullInfo$funebris, fullInfo$mercatorum, fullInfo$repleta, fullInfo$immigrans))
responseSpeciesComposition <- as.data.frame(lapply(responseSpeciesComposition, as.numeric))
responseSpeciesComposition[is.na(responseSpeciesComposition)] <- 0
responseSpeciesComposition <- subset(responseSpeciesComposition, select= -others)


ViennaClimate <- as.data.frame(ViennaData[4:ncol(ViennaData)])
rownames(ViennaClimate) <- ViennaData$sampleId
ViennaClimate_LN <- as.data.frame(ViennaData_LN[4:ncol(ViennaData_LN)])

ViennaClean <- ViennaClimate %>% select(where(~ any(. != 0)))
ViennaCleanLN <- ViennaClimate_LN %>% select(where(~ any(. != 0)))

pr_Vienna <- prcomp(ViennaClean, scale. = TRUE)
pr_ViennaLN <- prcomp(ViennaCleanLN, scale. = TRUE)

print("Data including the Land Usage Types")
fviz_pca_var(pr_ViennaLN, col.var = "contrib", gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"), repel = TRUE)
print("Land Usage Types removed from data")
fviz_pca_var(pr_Vienna, col.var = "contrib", gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"), repel = TRUE)
fviz_pca_biplot(pr_Vienna)
corrplot(cor(ViennaClean))


common_columns <- intersect(names(ViennaClean), names(fullInfo))
predictorVienna <- fullInfo[, common_columns]

rda_result <- rda(responseSpeciesComposition ~ ., data = predictorVienna)


##Vienna End

## Plotting Abundance Data

```

``` {r ViennaData, echo=FALSE}


##Species Count Total
df <- responseSpeciesComposition
plot_ly(
    x = colnames(df),
    y = colSums(df != 0),
    type = 'bar',
    marker = list(color = 'lightblue')
) %>%
    layout(title = 'Number of total Observations per Species',
           xaxis = list(title = 'Drosophila species'),
           yaxis = list(title = 'Species Count'))



## Missing Data 

binary_matrix <- ifelse(df == 0, 1, 0)

# Plot the matrix
image(t(binary_matrix), axes = FALSE, col = c("#4aa5d4", "#b3475f"))
axis(1, at = seq(0, 1, length.out = ncol(df)), labels = colnames(df), las = 2)
axis(2, at = seq(0, 1, length.out = nrow(df)), labels = 1:nrow(df))
title("Matrix Plot of Zero Values")

#par(mfrow = c(1, 2))
plot(rda_result)
ordiplot(rda_result, xlim = c(-20,20), ylim= c(-10,30))

#par(mfrow = c(1, 1))
``` 

``` {r Rasdaman_data ,fig.width=20, fig.height=14, echo=FALSE}
null_values <- read.csv("/media/inter/ssteindl/FC/usecaserepo/SYNC0524/uc3-drosophola-genetics/projects/WormPickerOOP/NUllValuesDetail.csv", header=FALSE)

null_values$V1 <- gsub("--7000--", "..7000..", null_values$V1)

# Nur Spaltennamen auswählen, die sowohl in null_values$V1 als auch in df_numeric vorhanden sind
valid_columns <- intersect(null_values$V1, colnames(df_numeric))

# Diese validen Spalten aus df_numeric auswählen
new <- df_numeric[, valid_columns]

filtered_null_values <- null_values[null_values$V1 %in% colnames(df_numeric),]
named_list <- as.list(setNames(filtered_null_values$V2, filtered_null_values$V1))

null_values_list <- named_list

df_new <- as.data.frame(lapply(seq_along(new), function(i) {
  replace(new[[i]], new[[i]] == null_values_list[[names(new)[i]]], NA)
}))

colnames(df_new) <- colnames(new)

colnames(df_new) <- gsub("ds.earthserver.xyz..7000..", "", colnames(df_new))
colnames(df_new) <- gsub("global_pesticide_grids_apr", "pest", colnames(df_new))
#C3S_satellite_soil_moisture
colnames(df_new) <- gsub("C3S_satellite_soil_moisture", "C3S_sm", colnames(df_new))
colnames(df_new) <- gsub("carbonmonoxide", "", colnames(df_new))
colnames(df_new) <- gsub("carbonmonoxide", "", colnames(df_new))


##find all-NA_columns


#na_columns <- names(df_new)[apply(df_new, 2, function(x) all(is.na(x)))]

#df_new_clean <- df_new[, !apply(df_new, 2, function(x) all(is.na(x)))]

rownames(df_new)<- Rasdman$sample.
core_europe <- df_new[!grepl("^UA|RU|BY|DK", rownames(df_new)), ]
na_count_col <- sapply(core_europe, function(x) sum(is.na(x)))
cols_to_keep <- na_count_col <= 0
remains <- core_europe[, cols_to_keep]

print(colnames(remains))

colnames(remains) <- paste(substr(colnames(remains), 1, 4), 1:ncol(remains), sep = "_")
pr_eur <- prcomp(remains, scale. = TRUE)




fviz_pca_var(pr_eur, col.var = "contrib", gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"), repel = TRUE)

```



## Data 
We have a lot of pesticde layers in our data, we do not print all 400 names of them. Non-Pesticide data that does not contain NAs throughout the whole samples are the following:

```{r pressure, echo=FALSE}

without_pesticides <- df_numeric[, !grepl("global", names(df_numeric))]
rownames(without_pesticides) <- Rasdman$sample.
rownames(without_pesticides) <- paste(substr(rownames(without_pesticides), 1, 2), 1:nrow(without_pesticides), sep = "_")
#rownames(without_pesticides) <- substr(rownames(without_pesticides), 1, nchar(rownames(without_pesticides)) - 10)
colnames(without_pesticides) <- gsub("ds.earthserver.xyz..7000..", "", colnames(without_pesticides))
#colnames(without_pesticides) <- gsub("global_pesticide_grids_apr", "pesticide", colnames(without_pesticides))
#C3S_satellite_soil_moisture
colnames(without_pesticides) <- gsub("C3S_satellite_soil_moisture", "C3S_sm", colnames(without_pesticides))
#colnames(without_pesticides)
# Identify columns with zero variance
zero_variance_columns <- sapply(without_pesticides, function(x) var(x, na.rm = TRUE) == 0)

# Remove columns with zero variance
without_pesticides_clean <- without_pesticides[, !zero_variance_columns]


cols_with_na <- apply(without_pesticides_clean, 2, function(x) any(is.na(x)))

# Remove columns with NA values
df_cleaned <- without_pesticides_clean[, !cols_with_na]

colnames(df_cleaned) <- paste(substr(colnames(df_cleaned), 1, 6), 1:ncol(df_cleaned), sep = "")
```

## PCA biplot on data without pesticide layers, non-NA data only

```{r pcaplot, fig.width=20, fig.height=14 , echo=FALSE}

pr <- prcomp(df_cleaned, scale. = TRUE)
fviz_pca_biplot(pr)

``` 


## Screeplots
```{r screeplot, echo=FALSE}

# Perform PCA on cleaned dataset
pca_result <- prcomp(df_cleaned, scale. = TRUE)
screeplot(pca_result, type = "lines")

```

## Biplot Colored

```{r variance_reduction,fig.width=20, fig.height=14, echo=FALSE}
fviz_pca_var(pca_result, col.var = "contrib", gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))



```



### Find influence of Variables


```{r Variable_Contributions, fig.width=20, fig.height=14,  echo=FALSE}
loadings <- pca_result$rotation

# Convert loadings to a data frame for ggplot
loadings_df <- as.data.frame(loadings[, 1])  # PC1 loadings
loadings_df$Variable <- rownames(loadings_df)
colnames(loadings_df) <- c("Loading", "Variable")

# Create a bar plot for PC1
ggplot(loadings_df, aes(x = reorder(Variable, Loading), y = Loading)) +
  geom_bar(stat = "identity") +
  coord_flip() +  # Flip coordinates for better readability
  theme_minimal() +
  labs(title = "Variable Contributions to PC1",
       x = "Variable",
       y = "Loading")
```

## Correlation of Environmental Factors 
``` {r correlation_plots,  fig.width=20, fig.height=14, echo =FALSE}

library(reshape2)
library(igraph)
library(qgraph)
# Melt the correlation matrix into long format
cor_matrix <- cor(df_cleaned)
melted_cor_matrix <- melt(cor_matrix)

# Plot with ggplot2
ggplot(data = melted_cor_matrix, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile() +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", midpoint = 0) +
  theme_minimal() +
  theme(axis.text.x = element_blank(), # Remove x-axis labels
        axis.title = element_blank())  # Remove axis titles


# Set a threshold to filter only strong correlations
cor_threshold <- 0.7  # Adjust based on your needs

colnames(cor_matrix) <- paste(substr(colnames(cor_matrix), 1, 3), 1:ncol(cor_matrix), sep = "")

# Keep only correlations above the threshold
strong_cor <- cor_matrix * (abs(cor_matrix) > cor_threshold)

# Plot a network graph of the strong correlations
qgraph(strong_cor, layout = "spring", vsize = 5,label.cex = 0.7, 
       edge.color = "gray", cut = cor_threshold, repulsion = 0.5, 
       title = "Network of Strong Correlations")
```
## Pesticides only 
``` {r pesticide_only, fig.width=20, fig.height=14, echo=FALSE}

pesticides <- df_numeric[, grepl("global_pesticide", names(df_numeric))]

colnames(pesticides) <- gsub("global_pesticide_grids_apr", "pc", colnames(pesticides))
rownames(pesticides) <- Rasdman$sample.
rownames(pesticides) <- paste(substr(rownames(pesticides), 1, 2), 1:nrow(pesticides), sep = "_")

pr <- prcomp(pesticides, scale. = TRUE)
fviz_pca_biplot(pr)
fviz_eig(pr, addlabels = TRUE) + theme_bw()

``` 



## Including Pesticide Data 

``` {r pesticide_inclusion, fig.width=20, fig.height=14, echo =FALSE}
library(pheatmap)
# Identify columns with zero variance
zero_variance_columns_pest <- sapply(df_numeric, function(x) var(x, na.rm = TRUE) == 0)

# Remove columns with zero variance
incl_pesticides_clean <- df_numeric[, !zero_variance_columns_pest]
cols_with_na <- apply(incl_pesticides_clean, 2, function(x) any(is.na(x)))

# Remove columns with NA values
pest_cleaned <- incl_pesticides_clean[, !cols_with_na]
rownames(pest_cleaned) <- Rasdman$sample.
rownames(pest_cleaned) <- paste(substr(rownames(pest_cleaned), 1, 2), 1:nrow(pest_cleaned), sep = "_")
colnames(pest_cleaned) <- gsub("ds.earthserver.xyz..7000..", "", colnames(pest_cleaned))
colnames(pest_cleaned) <- gsub("global_pesticide_grids_apr", "pesticide", colnames(pest_cleaned))
#C3S_satellite_soil_moisture
colnames(pest_cleaned) <- gsub("C3S_satellite_soil_moisture", "C3S_sm", colnames(pest_cleaned))

pest_r <- prcomp(pest_cleaned, scale. = TRUE)
#fviz_pca_var(pest_r, col.var = "contrib", gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))
fviz_pca_biplot(pest_r)

cor_matrix_pest <- cor(pest_cleaned)
cor_matrix <- cor(pest_cleaned, use = "pairwise.complete.obs")
library(pheatmap)
# Plot a clustered heatmap of the correlation matrix
pheatmap(cor_matrix, 
         clustering_distance_rows = "euclidean",  # Distance measure for rows (or columns)
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",          # Clustering method
         labels_col = NULL, # Adjust font size for large plots
         main = "Clustered Heatmap of Correlations")



glypho <- df_numeric[, grepl("Glyphosate", names(df_numeric))]
colnames(glypho) <- gsub("global_pesticide_grids_apr", "pc", colnames(glypho))
rownames(glypho) <- Rasdman$sample.
#rownames(glypho) <- paste(substr(rownames(glypho), 1, 2), 1:nrow(glypho), sep = "_")

pr <- prcomp(glypho, scale. = TRUE)
fviz_pca_biplot(pr)
fviz_eig(pr, addlabels = TRUE) + theme_bw()

```


#### Network Graph

```{r network, echo =FALSE}

library(igraph)
library(qgraph)



# Set a threshold to filter only strong correlations
cor_threshold <- 0.7  # Adjust based on your needs

# Keep only correlations above the threshold
strong_cor <- cor_matrix * (abs(cor_matrix) > cor_threshold)

# Plot a network graph of the strong correlations
qgraph(strong_cor, layout = "spring", vsize = 10,label.cex = 1.5, 
       edge.color = "gray", cut = cor_threshold, repulsion = 0.9, 
       title = "Network of Strong Correlations")
```


```{r RDAtest, echo =FALSE}
library(readr)
library(robust)
library(qvalue)
samples <- intersect(Rasdman$sample., metafile$sampleId)
AFs <- read_table2("/media/inter/ssteindl/FC/usecaserepo/SYNC0524/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FullData2/results/fullgenome/Subsampled_fullgenome.final_DP15.af")

sel <- intersect(rownames(remains), samples)
sel2 <- intersect(sel, colnames(AFs))

AlleleFreqs <- t(AFs[, sel2])
Env <- remains[sel2,]
Parameters <- colnames(Env)

RDA0 <- rda(AlleleFreqs ~ 1, Env)

formula_string <- paste("AlleleFreqs ~", paste(Parameters, collapse = " + "))
formula <- as.formula(formula_string)

RDAfull <- rda(formula, data = Env)

mod <- ordiR2step(RDA0, RDAfull, Pin = 0.01, R2permutations = 1000, R2scope = T)
Vars <- rownames(mod$anova)[1:(nrow(mod$anova)-1)]
#substr(rownames(mod$anova)[1],3,nchar(rownames(mod$anova)[1]))
#substr(Vars[1],3,nchar(Vars[1]))

for (i in 1:length(Vars)){
  Vars[i] <- substr(Vars[i],3,nchar(Vars[i]))
}

formula_string <- paste("AlleleFreqs ~", paste(Vars, collapse = " + "))
formula <- as.formula(formula_string)

###pRDAs?
#let us assume the outcome of the ordi2step is 5 variables:"TotalPrecipitation_tp.0"  "ERA5_d2m.0"  "dominant_leaf_type_10m_dlt.0"  

##we did not include neutral genetic strucutre yet

#RDA_env <- rda(AlleleFreqs ~ TotalPrecipitation_tp.0 + ERA5_d2m.0 + dominant_leaf_type_10m_dlt.0 ,  Env)
RDA_env <- rda(formula, data = Env)
screeplot(RDA_env, main="Eigenvalues of constrained axes")

source("/media/inter/ssteindl/FC/usecaserepo/SYNC0524/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/RDA_utils/src/rdadapt.R")

rdadapt_env<-rdadapt(RDA_env, 2)
Loci <- paste(AFs$Chr,AFs$Pos, sep=".")

## P-values threshold after Bonferroni correction
thres_env <- 0.01/length(rdadapt_env$p.values)
colnames(AlleleFreqs) <- Loci
AllFreq <- AlleleFreqs

## Identifying the loci that are below the p-value threshold
outliers <- data.frame(Loci = colnames(AllFreq)[which(rdadapt_env$p.values<thres_env)], p.value = rdadapt_env$p.values[which(rdadapt_env$p.values<thres_env)], contig = unlist(lapply(strsplit(colnames(AllFreq)[which(rdadapt_env$p.values<thres_env)], split = "_"), function(x) x[1])))


## Top hit outlier per contig
outliers <- outliers[order(outliers$contig, outliers$p.value),]

## List of outlier names
outliers_rdadapt_env <- as.character(outliers$Loci[!duplicated(outliers$contig)])

## Formatting table for ggplot
locus_scores <- scores(RDA_env, choices=c(1:2), display="species", scaling="none") # vegan references "species", here these are the loci
TAB_loci <- data.frame(names = row.names(locus_scores), locus_scores)
TAB_loci$type <- "Neutral"
TAB_loci$type[TAB_loci$names%in%outliers$Loci] <- "All outliers"
TAB_loci$type[TAB_loci$names%in%outliers_rdadapt_env] <- "Top outliers"
TAB_loci$type <- factor(TAB_loci$type, levels = c("Neutral", "All outliers", "Top outliers"))
TAB_loci <- TAB_loci[order(TAB_loci$type),]
TAB_var <- as.data.frame(scores(RDA_env, choices=c(1,2), display="bp")) # pull the biplot scores

ggplot() +
  geom_hline(yintercept=0, linetype="dashed", color = gray(.80), size=0.6) +
  geom_vline(xintercept=0, linetype="dashed", color = gray(.80), size=0.6) +
  geom_point(data = TAB_loci, aes(x=RDA1*20, y=RDA2*20, colour = type), size = 1.4) +
  scale_color_manual(values = c("gray90", "#F9A242FF", "#6B4596FF")) +
  geom_segment(data = TAB_var, aes(xend=RDA1, yend=RDA2, x=0, y=0), colour="black", size=0.15, linetype=1, arrow=arrow(length = unit(0.02, "npc"))) +
  geom_text(data = TAB_var, aes(x=1.1*RDA1, y=1.1*RDA2, label = row.names(TAB_var)), size = 2.5, family = "Times") +
  xlab("RDA 1") + ylab("RDA 2") +
  facet_wrap(~"RDA space") +
  guides(color=guide_legend(title="Locus type")) +
  theme_bw(base_size = 11, base_family = "Times") +
  theme(panel.background = element_blank(), legend.background = element_blank(), panel.grid = element_blank(), plot.background = element_blank(), legend.text=element_text(size=rel(.8)), strip.text = element_text(size=11))

```



