{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8c6c0820-66c6-4056-ab10-5cebc74dc92d",
   "metadata": {},
   "source": [
    "## **Landscape Genomics Pipeline**\n",
    "\n",
    "This Notebook runs a Landscape Genomic Analysis Pipeline on allele frequency data and environmental data.\n",
    "\n",
    "Default settings are set to run the following analysis:\\\n",
    "Full set of SNPs data for Drosophila melanogaster (DEST.bio) in European Population samples, analysed with WorldClim BIO1 and Copernicus Near Surface Air Temperature. \n",
    "\n",
    "**Parameters Required:**\n",
    "- Working Directory\n",
    "- Subset of Samples\n",
    "- Name of Project\n",
    "- Data Folder (Contains Environemntal Data and Genomic Data (downlaoded already))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a5d9d882-6119-40b5-bf78-28a83f93347c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2024-07-15 12:16:02.192\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36m__main__\u001b[0m:\u001b[36m<module>\u001b[0m:\u001b[36m11\u001b[0m - \u001b[1mProcessing data for year \u001b[0m\n",
      "\u001b[32m2024-07-15 12:16:02.215\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36m__main__\u001b[0m:\u001b[36m<module>\u001b[0m:\u001b[36m12\u001b[0m - \u001b[1mFinished saving data\u001b[0m\n",
      "\u001b[32m2024-07-15 12:16:02.216\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mmeasurer\u001b[0m:\u001b[36mstart\u001b[0m:\u001b[36m279\u001b[0m - \u001b[1mStarted computational costs meter: wall time, memory consumed, network traffic, CO2 emissions, data size\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.append('/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/ResourceLogging')\n",
    "\n",
    "import os\n",
    "import time\n",
    "from loguru import logger\n",
    "from measurer import Measurer\n",
    "from types import ModuleType\n",
    "\n",
    "logger.add(\"logfile.log\")\n",
    "logger.info(f\"Processing data for year \")\n",
    "logger.info(\"Finished saving data\")\n",
    "\n",
    "data_path = '/'\n",
    "measurer = Measurer()\n",
    "tracker = measurer.start(data_path=data_path, logger=logger)\n",
    "shape = []\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7ef190e-d591-4b8e-b4c3-e659beede917",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE’: File exists\n",
      "mkdir: cannot create directory ‘results’: File exists\n",
      "mkdir: cannot create directory ‘data’: File exists\n",
      "mkdir: cannot create directory ‘/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData’: File exists\n",
      "mkdir: cannot create directory ‘/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/Summary’: File exists\n",
      "DOWNLOADING VCF FROM DEST.bio AND EXTRACTING SAMPLENAMES\n",
      "DONE\n",
      "Starting VCFtools\n",
      "Finished VCFtools\n",
      "Starting LEA analysis\n",
      "mkdir: cannot create directory ‘/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA’: File exists\n",
      "sampleId,lat,long,nFlies,bio1,bio2,bio3,bio4,bio5,bio6,bio7,bio8,bio9,bio10,bio11,bio12,bio13,bio14,bio15,bio16,bio17,bio18,bio19\n",
      "lat\n",
      "\u001b[?25h\u001b[?25h── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.4     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "\u001b[?25h\u001b[?25h\u001b[?25hWarning message:\n",
      "In dir.create(repdir) :\n",
      "  '/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/lat' already exists\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"After removing monomorphic rows,526290rows are being analysed\"\n",
      "\u001b[?25h\u001b[?25h[1] \"File exists and has been opened.\"\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/lat/gradients.env\"\n",
      "\u001b[?25h\u001b[?25h[1] \"CALCULATING LFMM2\"\n",
      "\u001b[?25h[1] \"*********************************************************\"\n",
      "[1] \" Conversion from the lfmm format to the geno format\"\n",
      "[1] \"*********************************************************\"\n",
      "The project is saved into :\n",
      " results/FullData/LEA/genotypes.snmfProject \n",
      "\n",
      "To load the project, use:\n",
      " project = load.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n",
      "To remove the project, use:\n",
      " remove.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n",
      "[1] 240738410\n",
      "[1] \"*************************************\"\n",
      "[1] \"*          create.dataset            *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)                 296\n",
      "        -L (number of loci)                        526290\n",
      "        -s (seed random init)                      240738410\n",
      "        -r (percentage of masked data)             0.05\n",
      "        -x (genotype file in .geno format)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.geno\n",
      "        -o (output file in .geno format)           /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "\n",
      " Write genotype file with masked data, /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno:\t\tOK.\n",
      "\n",
      "[1] \"*************************************\"\n",
      "[1] \"* sNMF K = 1  repetition 1      *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)             296\n",
      "        -L (number of loci)                    526290\n",
      "        -K (number of ancestral pops)          1\n",
      "        -x (input file)                        /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "        -q (individual admixture file)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.Q\n",
      "        -g (ancestral frequencies file)        /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.G\n",
      "        -i (number max of iterations)          200\n",
      "        -a (regularization parameter)          10\n",
      "        -s (seed random init)                  4607182419040755818\n",
      "        -e (tolerance error)                   1E-05\n",
      "        -p (number of processes)               1\n",
      "        - diploid\n",
      "\n",
      "Read genotype file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno:\t\tOK.\n",
      "\n",
      "\n",
      "Main algorithm:\n",
      "\n",
      "Least-square error: 1168455.668531\n",
      "Write individual ancestry coefficient file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.Q:\t\tOK.\n",
      "Write ancestral allele frequency coefficient file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.G:\tOK.\n",
      "\n",
      "[1] \"*************************************\"\n",
      "[1] \"*    cross-entropy estimation       *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)         296\n",
      "        -L (number of loci)                526290\n",
      "        -K (number of ancestral pops)      1\n",
      "        -x (genotype file)                 /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.geno\n",
      "        -q (individual admixture)          /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.Q\n",
      "        -g (ancestral frequencies)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.G\n",
      "        -i (with masked genotypes)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "        - diploid\n",
      "\n",
      "Cross-Entropy (all data):\t 0.0126759\n",
      "Cross-Entropy (masked data):\t 0.0129981\n",
      "The project is saved into :\n",
      " results/FullData/LEA/genotypes.snmfProject \n",
      "\n",
      "To load the project, use:\n",
      " project = load.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n",
      "To remove the project, use:\n",
      " remove.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n",
      "[1] 1367686372\n",
      "[1] \"*************************************\"\n",
      "[1] \"*          create.dataset            *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)                 296\n",
      "        -L (number of loci)                        526290\n",
      "        -s (seed random init)                      1367686372\n",
      "        -r (percentage of masked data)             0.05\n",
      "        -x (genotype file in .geno format)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.geno\n",
      "        -o (output file in .geno format)           /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "\n",
      " Write genotype file with masked data, /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno:\t\tOK.\n",
      "\n",
      "[1] \"*************************************\"\n",
      "[1] \"* sNMF K = 1  repetition 2      *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)             296\n",
      "        -L (number of loci)                    526290\n",
      "        -K (number of ancestral pops)          1\n",
      "        -x (input file)                        /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "        -q (individual admixture file)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.Q\n",
      "        -g (ancestral frequencies file)        /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.G\n",
      "        -i (number max of iterations)          200\n",
      "        -a (regularization parameter)          10\n",
      "        -s (seed random init)                  4607182420167703780\n",
      "        -e (tolerance error)                   1E-05\n",
      "        -p (number of processes)               1\n",
      "        - diploid\n",
      "\n",
      "Read genotype file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno:\t\tOK.\n",
      "\n",
      "\n",
      "Main algorithm:\n",
      "\n",
      "Least-square error: 1168336.377941\n",
      "Write individual ancestry coefficient file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.Q:\t\tOK.\n",
      "Write ancestral allele frequency coefficient file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.G:\tOK.\n",
      "\n",
      "[1] \"*************************************\"\n",
      "[1] \"*    cross-entropy estimation       *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)         296\n",
      "        -L (number of loci)                526290\n",
      "        -K (number of ancestral pops)      1\n",
      "        -x (genotype file)                 /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.geno\n",
      "        -q (individual admixture)          /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.Q\n",
      "        -g (ancestral frequencies)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.G\n",
      "        -i (with masked genotypes)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "        - diploid\n",
      "\n",
      "Cross-Entropy (all data):\t 0.0126689\n",
      "Cross-Entropy (masked data):\t 0.0131334\n",
      "The project is saved into :\n",
      " results/FullData/LEA/genotypes.snmfProject \n",
      "\n",
      "To load the project, use:\n",
      " project = load.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n",
      "To remove the project, use:\n",
      " remove.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n",
      "\u001b[?25h[1] \"TESTING THE LFMM2 OBJECT\"\n",
      "\u001b[?25h\u001b[?25h[1] \"EXPORTIN P-VALUES TO CSV FILE\"\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"PLOT MANHATTAN\"\n",
      "\u001b[?25h\u001b[?25h\u001b[?25hNULL\n",
      "\u001b[?25hError in plot_list.append(pl) : \n",
      "  could not find function \"plot_list.append\"\n",
      "Execution halted\n",
      "\u001b[?25hlong\n",
      "\u001b[?25h\u001b[?25h── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.4     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "\u001b[?25h\u001b[?25h\u001b[?25hWarning message:\n",
      "In dir.create(repdir) :\n",
      "  '/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/long' already exists\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"After removing monomorphic rows,526290rows are being analysed\"\n",
      "\u001b[?25h\u001b[?25h[1] \"File exists and has been opened.\"\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/long/gradients.env\"\n",
      "\u001b[?25h\u001b[?25h[1] \"CALCULATING LFMM2\"\n",
      "\u001b[?25h[1] \"*********************************************************\"\n",
      "[1] \" Conversion from the lfmm format to the geno format\"\n",
      "[1] \"*********************************************************\"\n",
      "The project is saved into :\n",
      " results/FullData/LEA/genotypes.snmfProject \n",
      "\n",
      "To load the project, use:\n",
      " project = load.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n",
      "To remove the project, use:\n",
      " remove.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n",
      "[1] 443605693\n",
      "[1] \"*************************************\"\n",
      "[1] \"*          create.dataset            *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)                 296\n",
      "        -L (number of loci)                        526290\n",
      "        -s (seed random init)                      443605693\n",
      "        -r (percentage of masked data)             0.05\n",
      "        -x (genotype file in .geno format)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.geno\n",
      "        -o (output file in .geno format)           /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "\n",
      " Write genotype file with masked data, /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno:\t\tOK.\n",
      "\n",
      "[1] \"*************************************\"\n",
      "[1] \"* sNMF K = 1  repetition 1      *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)             296\n",
      "        -L (number of loci)                    526290\n",
      "        -K (number of ancestral pops)          1\n",
      "        -x (input file)                        /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "        -q (individual admixture file)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.Q\n",
      "        -g (ancestral frequencies file)        /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.G\n",
      "        -i (number max of iterations)          200\n",
      "        -a (regularization parameter)          10\n",
      "        -s (seed random init)                  3328195528485823165\n",
      "        -e (tolerance error)                   1E-05\n",
      "        -p (number of processes)               1\n",
      "        - diploid\n",
      "\n",
      "Read genotype file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno:\t\tOK.\n",
      "\n",
      "\n",
      "Main algorithm:\n",
      "\n",
      "Least-square error: 1167747.148272\n",
      "Write individual ancestry coefficient file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.Q:\t\tOK.\n",
      "Write ancestral allele frequency coefficient file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.G:\tOK.\n",
      "\n",
      "[1] \"*************************************\"\n",
      "[1] \"*    cross-entropy estimation       *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)         296\n",
      "        -L (number of loci)                526290\n",
      "        -K (number of ancestral pops)      1\n",
      "        -x (genotype file)                 /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.geno\n",
      "        -q (individual admixture)          /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.Q\n",
      "        -g (ancestral frequencies)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run1/genotypes_r1.1.G\n",
      "        -i (with masked genotypes)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "        - diploid\n",
      "\n",
      "Cross-Entropy (all data):\t 0.012671\n",
      "Cross-Entropy (masked data):\t 0.0130919\n",
      "The project is saved into :\n",
      " results/FullData/LEA/genotypes.snmfProject \n",
      "\n",
      "To load the project, use:\n",
      " project = load.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n",
      "To remove the project, use:\n",
      " remove.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n",
      "[1] 999197623\n",
      "[1] \"*************************************\"\n",
      "[1] \"*          create.dataset            *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)                 296\n",
      "        -L (number of loci)                        526290\n",
      "        -s (seed random init)                      999197623\n",
      "        -r (percentage of masked data)             0.05\n",
      "        -x (genotype file in .geno format)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.geno\n",
      "        -o (output file in .geno format)           /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "\n",
      " Write genotype file with masked data, /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno:\t\tOK.\n",
      "\n",
      "[1] \"*************************************\"\n",
      "[1] \"* sNMF K = 1  repetition 2      *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)             296\n",
      "        -L (number of loci)                    526290\n",
      "        -K (number of ancestral pops)          1\n",
      "        -x (input file)                        /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "        -q (individual admixture file)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.Q\n",
      "        -g (ancestral frequencies file)        /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.G\n",
      "        -i (number max of iterations)          200\n",
      "        -a (regularization parameter)          10\n",
      "        -s (seed random init)                  4607182419799215031\n",
      "        -e (tolerance error)                   1E-05\n",
      "        -p (number of processes)               1\n",
      "        - diploid\n",
      "\n",
      "Read genotype file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno:\t\tOK.\n",
      "\n",
      "\n",
      "Main algorithm:\n",
      "\n",
      "Least-square error: 1168359.208887\n",
      "Write individual ancestry coefficient file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.Q:\t\tOK.\n",
      "Write ancestral allele frequency coefficient file /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.G:\tOK.\n",
      "\n",
      "[1] \"*************************************\"\n",
      "[1] \"*    cross-entropy estimation       *\"\n",
      "[1] \"*************************************\"\n",
      "summary of the options:\n",
      "\n",
      "        -n (number of individuals)         296\n",
      "        -L (number of loci)                526290\n",
      "        -K (number of ancestral pops)      1\n",
      "        -x (genotype file)                 /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.geno\n",
      "        -q (individual admixture)          /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.Q\n",
      "        -g (ancestral frequencies)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/K1/run2/genotypes_r2.1.G\n",
      "        -i (with masked genotypes)         /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/results/FullData/LEA/genotypes.snmf/masked/genotypes_I.geno\n",
      "        - diploid\n",
      "\n",
      "Cross-Entropy (all data):\t 0.0126753\n",
      "Cross-Entropy (masked data):\t 0.0129992\n",
      "The project is saved into :\n",
      " results/FullData/LEA/genotypes.snmfProject \n",
      "\n",
      "To load the project, use:\n",
      " project = load.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n",
      "To remove the project, use:\n",
      " remove.snmfProject(\"results/FullData/LEA/genotypes.snmfProject\")\n",
      "\n"
     ]
    }
   ],
   "source": [
    "WorkingDirectory=\"/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE\"\n",
    "Continent=\"Europe\"\n",
    "ProjectName=\"FullData\"\n",
    "ResourceLogger=\"/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/ResourceLogging/Resources_PipelineRun_Optimizations3.1.csv\"\n",
    "\n",
    "GenomicData=\"/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/data/PoolSeq2023.vcf.gz\"\n",
    "GenomicMetaData=\"/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/dest_v2.samps_3May2024.csv\"\n",
    "EnvironmentalData=\"/home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/data/metadata_new.csv\"\n",
    "GFF=\"s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/FULL_PIPELINE/data/dmel-all-r6.57.gff.gz\"\n",
    "\n",
    "!bash /home/sonjastndl/s3/LGA/uc3-drosophola-genetics/projects/LandscapeGenomicsPipeline/scripts/Landscape_Genomics.sh $WorkingDirectory $Continent $ProjectName $GenomicData $GenomicMetaData $EnvironmentalData\n",
    "\n",
    "measurer.end(tracker=tracker,\n",
    "             shape=shape,\n",
    "             libraries=[k for k, v in globals().items() if type(v) is ModuleType and not k.startswith('__')],\n",
    "             data_path=data_path,\n",
    "             variables=locals(),\n",
    "             csv_file=ResourceLogger,\n",
    "            logger=logger)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff9421a1-081f-404b-95fd-64fb37c126fe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "fairicube-edc-2022.10-14",
   "language": "python",
   "name": "conda-env-fairicube-edc-2022.10-14-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
